File: neural_network/components/layer.py
from typing import List

from .neuron import Neuron


class Layer:
    """Class to represent one `Layer` of a network
    """

    def __init__(self, _id: int, num_neurons: int):
        """Constructor method

        Parameters
        ----------
        _id : int
            The id of the `Layer`
        num_neurons : int
            The number of `Neurons` in the `Layer`
        """
        self._id = _id
        self._num_neurons = num_neurons
        self._neurons = [Neuron(_id, j) for j in range(num_neurons)]

    def get_id(self) -> int:
        """Getter method

        Returns
        -------
        int
            The id
        """
        return self._id

    def get_neurons(self) -> List[Neuron]:
        """Getter method

        Returns
        -------
        List[Neuron]
            The `Neurons` in the `Layer`
        """
        return self._neurons

    def __str__(self):
        return f"Layer {self._id}"

File: javaNeuralNetwork/src/main/java/neural_network/components/Layer.java
package neural_network.components;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Stream;

/** Class to represent one {@code Layer} of a network.
 * @version 1.0.0
 * @since 1.0.0
 */
public class Layer {

    private final int id;
    private final int numNeurons;
    private final List<Neuron> neurons = new ArrayList<>();

    /** Constructor method.
     *
     * @param id The id of the {@code Layer}.
     * @param numNeurons The number of {@code Neurons} in the {@code Layer}.
     */
    Layer(int id, int numNeurons) {
        this.id = id;
        this.numNeurons = numNeurons;
        this.neurons.addAll(Stream
                .iterate(0, j -> j < numNeurons, j -> j + 1)
                .map(j -> new Neuron(id, j))
                .toList());
    }

    /** Getter method for {@code id}.
     *
     * @return The {@code id} of the {@code Layer}.
     */
    int getId() {
        return id;
    }

    /** Getter method for {@code neurons}
     *
     * @return A {@code List} of {@code Neurons} in the {@code Layer}.
     */
    List<Neuron> getNeurons() {
        return List.copyOf(neurons);
    }

    /** The size of the {@code Layer}.
     *
     * @return The number of {@code Neurons} in the {@code Layer}.
     */
    int size() {
        return numNeurons;
    }

    /** String method.
     *
     * @return {@code String} representation of the {@code Layer}.
     */
    @Override
    public String toString() {
        return "Layer %d".formatted(id);
    }
}
